rewrite Rust (hyprlang-rs)

RAprogramm opened the pull request:
Closes #2

Commit:
Initiate hyprQuery

kRHYME7 committed:
Create cmake-single-platform.yml

Commit:
yea

Commit:
static build hyprlang so we can redistribute

Commit:
hyprutils? I'm DUUMB!!!

Commit:
Add release

Commit:
v0.0.2

kRHYME7 committed:
working $variable qeury

kRHYME7 committed:
use wordexp

kRHYME7 committed:
gracefully handle filepath errors

kRHYME7 committed:
output none on plain txt instead of sring:NULL

kRHYME7 committed:
feat: dynamic lin installation

kRHYME7 committed:
fix

kRHYME7 committed:
Merge branch 'main' of https://github.com/kRHYME7/hyprquery

kRHYME7 committed:
fixe

kRHYME7 committed:
almost workign source imple

kRHYME7 committed:
source HAndler works now! FFFFFFFFFFFFFFFFFFFFFFFFFF

kRHYME7 committed:
modular

kRHYME7 committed:
modular files

kRHYME7 committed:
r me

kRHYME7 committed:
better path handler

kRHYME7 committed:
* Use static hyprlang
* Add clang and clang-tidy

kRHYME7 committed:
readme

kRHYME7 committed:
refactor: Removes the var lookup work around

kRHYME7 committed:
fix: Handle multi query

kRHYME7 committed:
build

kRHYME7 committed:
Update cmake-single-platform.yml

kRHYME7 committed:
Update cmake-single-platform.yml

kRHYME7 committed:
Update cmake-single-platform.yml add hyprutils build

kRHYME7 committed:
Update cmake-single-platform.yml add pixman

kRHYME7 committed:
Update cmake-single-platform.yml

kRHYME7 committed:
Update cmake-single-platform.yml

kRHYME7 committed:
Update cmake-single-platform.yml

kRHYME7 committed:
Release

kRHYME7 committed:
readme

kRHYME7 committed:
optionto expect type and regex

kRHYME7 committed:
Use stream to inject Dynamic Variables for look u p

kRHYME7 committed:
Feat: Added --query "query[type][regex]" - type and regex hints!

kRHYME7 committed:
Feat: Added --query "query[type][regex]" - type and regex hints!

kRHYME7 committed:
build spdlog log statically to avoid breakage.

RAprogramm committed:
#2 feat: rewrite hyprquery in Rust using hyprlang-rs

RAprogramm committed:
#2 feat: add unit tests and improve functionality

RAprogramm committed:
#2 refactor: split main.rs into logical modules

RAprogramm committed:
#2 refactor: move app logic to app.rs, keep only main fn in main.rs

RAprogramm committed:
#2 perf: optimize memory usage with Box<str> and static str

RAprogramm committed:
#2 docs: add comprehensive module and function documentation

RAprogramm committed:
#2 refactor: modernize error handling with masterror builder API

RAprogramm committed:
#2 feat: add colorful custom help command with detailed documentation

RAprogramm committed:
Rename project to hydequery

RAprogramm committed:
Fix help header to Hydequery

RAprogramm committed:
Add HyDE banner to README

RAprogramm committed:
#2 ci: add professional CI workflow with caching

RAprogramm committed:
#2 docs: add CI and coverage badges to README

kRHYME7 commented:
@prime-run 🙏🙏🙏

RAprogramm committed:
#2 feat: add schema fetching and caching with --fetch-schema and --schema auto

RAprogramm committed:
#2 format

kRHYME7 commented on README.md:
```diff
@@ -1,176 +1,271 @@
-A command-line utility for querying configuration values from Hyprland and hyprland-related configuration files using the hyprlang parsing library.
+# Hydequery
+
+[![CI](https://github.com/HyDE-Project/hyprquery/actions/workflows/ci.yml/badge.svg)](https://github.com/HyDE-Project/hyprquery/actions/workflows/ci.yml)
```
It is called hyprquery.

RAprogramm committed:
#2 docs: add comprehensive man page

RAprogramm committed:
#2 fix: hide invalid regex from clippy static analysis

RAprogramm committed:
#2 up ureq version

RAprogramm commented:
@kRHYME7 we need codecov token in org secrets!

kRHYME7 commented:
Please  test against 

<img width="2932" height="1108" alt="image" src="https://github.com/user-attachments/assets/8c94bf8b-39f5-40ee-a04a-6fbce10cc595" />

RAprogramm commented:
> Please test against
> <img alt="image" width="2000" height="1108" src="https://private-user-images.githubusercontent.com/53417443/517684749-8c94bf8b-39f5-40ee-a04a-6fbce10cc595.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM3OTY1MzksIm5iZiI6MTc2Mzc5NjIzOSwicGF0aCI6Ii81MzQxNzQ0My81MTc2ODQ3NDktOGM5NGJmOGItMzlmNS00MGVlLWEwNGEtNmZiY2UxMGNjNTk1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTIyVDA3MjM1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMxNjFjNzM4MTU2M2E4Yjk0MzIwNjIxMzk0M2Q2NGQzOWIzMDllYTFlZjBiY2YwOTRjZmE4ZWYzMmI0ZDRhMGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ENq0BJ2ug8zQORat5ZY-fQ_0gO5p0Ok2D2vdGEJJ6Ck">

give me more context, please, bro

kRHYME7 commented:
Oh, to avoid breakage please test it against the existing scripts that utilize HyDE.

RAprogramm committed:
#2 fix: strip comments from source directive paths

Source paths with inline comments were not parsed correctly,
causing the -s flag to fail when resolving paths like:
source = $HOME/.local/share/hyde/hyprland.conf # comment

Also improved error handling to skip unresolvable paths
instead of failing the entire source parsing.

RAprogramm commented:
Tested against HyDE scripts compatibility.

**Fixed issue**: Source paths with inline comments were not parsed correctly, causing the `-s` flag to fail. Example:
```
source = $HOME/.local/share/hyde/hyprland.conf # Fallback
```

**Tested all variables from HyDE `configuration.py`**:

```bash
$ hydequery ~/.config/hypr/hyprland.conf -s \
  -Q '$GTK_THEME' \
  -Q '$ICON_THEME' \
  -Q '$CURSOR_THEME' \
  -Q '$CURSOR_SIZE' \
  -Q '$COLOR_SCHEME'

Wallbash-Gtk
Tela-circle-dracula
Bibata-Modern-Ice
24
prefer-dark
```

**Export formats work correctly**:

env format (for shell scripts):
```bash
$ hydequery ... --export env
GTK_THEME="Wallbash-Gtk"
ICON_THEME="Tela-circle-dracula"
```

**Nested config values**:
```bash
$ hydequery ~/.config/hypr/hyprland.conf -s \
  -Q 'general:border_size' -Q 'decoration:rounding'
2
3
```

All HyDE theme variables are accessible through the main hyprland.conf with `-s` flag. Ready for integration with `configuration.py`.

RAprogramm committed:
#2 test: improve test coverage to 74%

Added comprehensive tests with real HyDE config examples:
- export.rs: 100% coverage
- source.rs: 96% coverage
- value.rs: 98% coverage
- query.rs: 97% coverage
- schema.rs: 96% coverage
- path.rs: 96% coverage

Integration tests cover real HyDE scenarios:
- Theme variables (GTK_THEME, ICON_THEME, CURSOR_SIZE)
- General settings (gaps, border_size, rounding)
- Source directive chaining
- Decoration and animation configs

RAprogramm committed:
#2 refactor: extract run_with_args for testability

Separated Args::parse() from main logic to enable comprehensive testing.

Coverage improved from 74% to 86%:
- app.rs: 93% (was 44%)
- Core business logic fully covered

Added 30+ tests for run_with_args covering:
- Theme variables (GTK_THEME, ICON_THEME, CURSOR_SIZE)
- General/decoration settings
- Source directive following
- Type and regex filters
- Error cases (missing vars, invalid paths)
- Export formats (json, env, plain)

RAprogramm committed:
#2 refactor: extract filters and defaults modules from app

RAprogramm committed:
#2 test: add tests for help module

RAprogramm committed:
#2 test: add tests for main, fetch and integration

RAprogramm committed:
#2 fix: make CLI more professional with proper error messages

- Make config_file optional to allow --fetch-schema standalone
- Add validation_error for clear user feedback
- Fix error display to show actual message instead of error kind
- Update help and man page to reflect new CLI patterns

RAprogramm committed:
#2 docs: update ASCII art to HYDEQUERY and improve README

- Fix ASCII art to spell HYDEQUERY instead of HYPRQUE
- Add filters.rs and defaults.rs to architecture section
- Add --get-defaults example
- Clarify that -Q is not required for all operations

RAprogramm committed:
#2 feat: add criterion benchmarks for performance testing

- Add benchmarks for query parsing (simple, typed, regex)
- Add benchmarks for file I/O operations
- Add benchmarks for full query execution
- Add lib.rs to expose modules for benchmarking
- Add Clone derive to Args for benchmark iterations
- Fix doctests in fetch.rs

RAprogramm commented:
- Query parsing: ~40-100ns
- File I/O: ~4µs

RAprogramm committed:
#2 feat: add automatic changelog generation with git-cliff

- Add cliff.toml configuration for conventional commits
- Add changelog CI job that runs on push to main
- Generate initial CHANGELOG.md from commit history
- Configure commit message parsing for features, fixes, docs, etc.

RAprogramm committed:
#2 feat: add Dependabot with grouped updates and auto-merge

- Configure weekly dependency checks for Cargo and GitHub Actions
- Group all minor/patch updates into single PRs by ecosystem
- Auto-merge and auto-approve minor/patch updates after CI passes
- Use squash merge for clean history

RAprogramm commented:
@kRHYME7, i'm think i'm done) let's check this stuff!

i'll add benchmark section in readme later

kRHYME7 commented:
@RAprogramm can you name the binary `hyq` as analogous to jq and yq ? also call it `hyprquery` please. Thank you!

RAprogramm commented:
@kRHYME7 works good?

RAprogramm commented:
> @RAprogramm can you name the binary `hyq` as analogous to jq and yq ? also call it `hyprquery` please. Thank you! 

So how it could be?
Hyq instead of hydequery?
Where I have to change?

kRHYME7 commented:
hyq for the binary executable name.
hyprquery for the whole package name.

hyprquery because it can be a stand alone tool to parse hypr configs.

RAprogramm commented:
Got you!

prime-run commented:
@RAprogramm would you mind pushing this against an empty `experimental/dev` branch ? 
would be way easier to colab  on.

edit: jsut for the sake of diffing rs files against rs file!

RAprogramm commented:
> @RAprogramm would you mind pushing this against an empty `experimental/dev` branch ? would be way easier to colab on.
> 
> edit: jsut for the sake of diffing rs files against rs file!

@prime-run dev branch created, bro! yes, it's a good practice to protect main/master branch

RAprogramm committed:
#2 refactor: rename package to hyprquery with hyq binary

RAprogramm commented:
@kRHYME7 , done, renamed to `hyq` and `hyprquery`

RAprogramm committed:
#2 ci: run checks on all PRs and update artifact name

prime-run commented:
@RAprogramm

cool, I clean up dev branch. would be great if you deal with conflicts. then we merge and take from there.

P.S maybe keep that warning I added on top of the README.md ?!

edit: let me squash everything in dev too.

prime-run commented:
@RAprogramm Alright, we have a clean history under `dev` now. I force pushed from an orphan branch so github automatically closed this PR.
You can fetch and cherry pick your commits and force push against dev (or just copy files over and commit all at once XD up to you ).
Sorry for the trouble but I think this is the cleanest way.

prime-run commented:
@RAprogramm I did the cherry picking and pushed to [hdev branch](https://github.com/HyDE-Project/hyprquery/commits/hdev/)

Would be great if you could confirm your commit history there.
